lib/DiscoveryService.php

```diff
@@ -0,0 +1,192 @@
+
+		// Read the memcached value (if the memcache is installed)
+		$discovery = $this->cache->get($discoveryKey);
+
+		if ($discovery === null) {
+			$this->logger->debug('Fetching {discoveryKey} as not found in cache', ['app' => 'richdocuments', 'discoveryKey' => $discoveryKey]);
+
+			try {
+				// If we are sending query to built-in CODE server, we avoid using IClient::get() method
+				// because of an encoding issue in guzzle: https://github.com/guzzle/guzzle/issues/1758
+				if (\strpos($wopiDiscovery, 'proxy.php') === false) {
+					$wopiClient = $this->httpClient->newClient();
+					$discovery = $wopiClient->get($wopiDiscovery)->getBody();
+				} else {
+					$discovery = \file_get_contents($wopiDiscovery);
+				}
+			} catch (\Exception $e) {
+				$error_message = $e->getMessage();
+
+				if (\preg_match('/^cURL error ([0-9]*):/', $error_message, $matches)) {
+					$curl_error = $matches[1];
+					switch ($curl_error) {
+						case '1':
+							$this->logger->error('Collabora Online: The protocol specified in {wopiRemote} is not allowed.', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+							break;
+						case '3':
+							$this->logger->error('Collabora Online: Malformed URL {wopiRemote}.', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+							break;
+						case '6':
+							$this->logger->error('Collabora Online: Cannot resolve the host {wopiRemote}.', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+							break;
+						case '7':
+							$this->logger->error('Collabora Online: Cannot connect to the host {wopiRemote}.', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+							break;
+						case '35':
+							$this->logger->error('Collabora Online: SSL/TLS handshake failed with the host {wopiRemote}.', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+							break;
+						case '60':
+							$this->logger->error('Collabora Online: SSL certificate is not installed. Please ask your administrator to add ca-chain.cert.pem to the ca-bundle.crt, for example "cat /etc/loolwsd/ca-chain.cert.pem >> <server-installation>/resources/config/ca-bundle.crt". The exact error message was: {error_message}', ['app' => 'richdocuments', 'error_message' => $error_message]);
+							break;
+						default:
+							$this->logger->error('Collabora Online cURL error: {error}', ['app' => 'richdocuments', 'error' => $error_message]);
+							break;
+					}
+				} else {
+					$this->logger->error('Collabora Online unknown error: {error}', ['app' => 'richdocuments', 'error' => $error_message]);
+				}
+
+				return null;
+			}
+
+			if (!$discovery) {
+				$this->logger->error('Collabora Online: Unable to read discovery.xml from {wopiRemote}', ['app' => 'richdocuments', 'wopiRemote' => $wopiRemote]);
+				return null;
+			}
+
+			$this->logger->debug('Storing {discoveryKey} to the cache.', ['app' => 'richdocuments', 'discoveryKey' => $discoveryKey]);
+			$this->cache->set($discoveryKey, $discovery, 3600);
```

Comment:
just to confirm, we don't need any kind of encryption nor encoding for the stored value, do we?